CIPs/0008.md

```diff
@@ -0,0 +1,76 @@
+
+Examples include:
+- Transaction metadata such as comments or encrypted messaging in general
+- Profile information such as name or a profile picture
+- Unidirectional attestations like an audit or badges won in The Great Celo Stake Off
+- Signed statements/claims by a party for relase binaries/docker images/package hashes
+
+## Goals
+
+- Expand storage and indexing capabilities of this off-chain data
+- Encode the notion of decryption keys
+- Lightweight messaging system
+
+
+## Proposed Solution
+
+We propose to add another claim, called the `STORAGE` claim:
+```json
+{
+  "type": "STORAGE",
+  "timestamp": 1588886114907,
+  "address": "https://example.com/celo_storage"
+}
+```
+
+Alternatively, this could be set directly on the `Accounts` smart contract directly, causing off-chain storage to effectively replace the existing metadata functionality. This would have the benefit of not duplicating similar functionality.
+
+Either way, this http address should be referred to as the root. The expectation here is that data can accessed via `root/${dataAddress}`. For private data, this should be encrypted to the recipient. Here are some common ways of modeling applications into this format:
+
+- For transaction metadata, the account can encrypt the metadata and put it in a file with the transaction hash as the metadata. Further CIPs could specify a standard for thee structure of that metadata.
+    - `dataAddress`: `transactionHash`
+    - file contents: Encrypted, structured transaction metadata (comments, or otherwise)
+- Data in relation to other accounts could be communicated to other accounts with:
+    - `dataAddress`: `otherAccountAddress`
+    - file contents: Structured data such as:
+        - `sessionKeys`: The session key(s) encrypted with dataEncryptionKey of `otherAccount` on `Accounts.sol`
+        - `attestations`: These could be signed data attestating things such an audit or badges. An attested account could also store the attestations from the attestor here, as a client can just verify the signature of the attestor.
+- Profile information could be stored on this system with:
+    - `dataAddress`: `profileName` or `profilePicture`
+    - file contents: the data unencrypted, or encrypted to session keys available to the recipients
+- Releases:
+    - `dataAddress`: `hashOfRelease`
+    - file contents: The signature of the key on the hash of the release
+
+If no assocation between to the `dataAddress` should be leaked, the `dataAddress` itself could be encrypted. The recipient would then just have to decrypt all `dataAddress`s (which could be available in an index file).
+
+All file contents should always be signed by an authorized signer of the account to prevent tampering by the file hoster.
```

Comment:
One alternative to specifying a signature to be included would be include a hash of the storage. If a directory tree structure is to be supported, it could be a hash tree of which the root is stored in the metadata entry.

Reply:
What are the advantages of your suggestion?

Reply:
Given the root hash for a storage tree at a given point in time, it would allow verification of the storage without needing to sign every piece of data in it. It would also make it impossible for someone to add or remove data from storage, including via MitM attack on the requestor, or someone compromising the storage server without access to the key that signs the metadata. Of course, it would require the root hash to be updated everytime a piece of data is added or removed, so it may be too cumbersome to use.

Reply:
Yeah it seems prohibitive to me to update the root hash for every change, what do you think?